drivers: uhc: use correct endpoint type and interval

AidenHu opened the pull request:
For usb xfer, set endpoint type and interval by the selected endpoint desc.

AidenHu committed:
drivers: uhc: use correct endpoint type and interval

In the USB transfer allocation, set endpoint type and interval
from the appropriate endpoint descriptor.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu committed:
drivers: uhc: add mcux_eps_interval to save ep interval

mcux_eps_interval is added as the new member of
uhc_mcux_data. It is used to save endpoint's
original interval value and can be compared
with xfer->interval.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

tmon-nordic commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -243,10 +243,9 @@ static usb_host_pipe_handle uhc_mcux_check_hal_ep(const struct device *dev,
+	if (mcux_ep != NULL && mcux_ep->pipeType == xfer->type &&
 	    (mcux_ep->maxPacketSize != xfer->mps ||
-	     mcux_ep->interval != xfer->interval)) {
+	     mcux_ep->interval != (uint16_t)(1UL << (xfer->interval - 1U)))) {
```
I don't know what `mcux_ep` expects in its `interval` field, but from USB specification point of view bInterval is not always used as exponent. For full-/low-speed interrupt endpoints, bInterval is value from 1 to 255.

AidenHu replied:
Sorry for the mistake, I have done based on the spec.

AidenHu committed:
drivers: uhc: set right value for pipe by xfer's mps

maxPacketSize and numberPerUframe of pipe should
be set considering additional transactions.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

MarkWangChinese commented on drivers/usb/uhc/uhc_common.c:
Add:
```
interval = 0;
type = USB_EP_TYPE_CONTROL;
```

MarkWangChinese commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -243,10 +284,21 @@ static usb_host_pipe_handle uhc_mcux_check_hal_ep(const struct device *dev,
+		actual_interval = (uint16_t)(1UL << (xfer->interval - 1U));
+	} else {
+		actual_interval = xfer->interval;
+	}
```
Suggest to add one struct member in `struct uhc_mcux_data` as follow. `mcux_ep->interval` is different with the interval value of endpoint descriptor.
```
	usb_host_pipe_t *mcux_eps[USB_HOST_CONFIG_MAX_PIPES];
    uint16_t mcux_eps_interval[USB_HOST_CONFIG_MAX_PIPES];
```

MarkWangChinese reviewed (commented):
The second commit body description needs updating.

AidenHu commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -235,17 +235,17 @@ static usb_host_pipe_handle uhc_mcux_check_hal_ep(const struct device *dev,
-	    priv->mcux_eps_interval[i] != xfer->interval)) {
+		(mcux_ep->maxPacketSize != xfer->mps ||
+		priv->mcux_eps_interval[i] != xfer->interval)) {
 		/* re-initialize the ep */
```
fixed

josuah reviewed (approved):
- Directly on top of main, so can be merged without dependency
- It adds a new property, without breaking existing drivers
- Using bInterval and bEndpointType from the descriptors or default values for EP0
- Add support for one particular driver

All looks good to me!
